internal/app/machined/pkg/controllers/block/internal/sysblock/sysblock.go

```diff
@@ -0,0 +1,144 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package sysblock implements gathering block device information from /sys/block filesystem.
+package sysblock
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/mdlayher/kobject"
+)
+
+// Walk the /sys/block filesystem and gather block device information.
+//
+//nolint:gocyclo
+func Walk(root string) ([]*kobject.Event, error) {
+	entries, err := os.ReadDir(root)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read %q: %w", root, err)
+	}
+
+	result := make([]*kobject.Event, 0, len(entries))
+
+	for _, entry := range entries {
+		fi, err := entry.Info()
+		if err != nil {
+			if os.IsNotExist(err) {
+				continue
+			}
+
+			return nil, fmt.Errorf("failed to stat %s: %w", entry.Name(), err)
+		}
+
+		if fi.Mode()&os.ModeSymlink == 0 {
+			continue
+		}
+
+		path, err := filepath.EvalSymlinks(filepath.Join(root, entry.Name()))
+		if err != nil {
+			if os.IsNotExist(err) {
+				continue
+			}
+
+			return nil, fmt.Errorf("failed to resolve symlink %s: %w", entry.Name(), err)
+		}
+
+		uevent, err := readUevent(path)
+		if err != nil {
+			if os.IsNotExist(err) {
+				continue
+			}
+
+			return nil, err
+		}
+
+		result = append(result, &kobject.Event{
+			Action:     kobject.Add,
```

Comment:
is this always an `Add` event?

Reply:
yes, as we're injecting current contents